File: solutions/860/main.py
from typing import List 

class Solution:
    def lemonadeChange(self, bills: List[int]) -> bool:
        numFive = 0
        numTen = 0
        for i in range(len(bills)):
            if bills[i] == 5:
                numFive += 1
            elif bills[i] == 10:
                numFive -= 1
                numTen += 1
                if numFive < 0:
                    return False
            else:
                if numTen > 0:
                    numTen -= 1
                    numFive -= 1
                    if numTen < 0 or numFive < 0:
                        return False
                else:
                    numFive -= 3
                    if numFive < 0:
                        return False
        return True

if __name__ == "__main__":
    mySol = Solution()
    bills = [5,5,5,10,20]
    assert mySol.lemonadeChange(bills)

File: solutions/860/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Sun Nov  7 19:29:17 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    bool lemonadeChange(vector<int>& bills) {
        int numFive = 0;
        int numTen = 0;
        for (auto& b : bills) {
            if (b == 5) numFive++;
            else if (b == 10) {
                numFive--;
                numTen++;
                if (numFive < 0) return false;
            }
            else {
                if (numTen > 0) {
                    numTen--;
                    numFive--;
                }
                else {
                    numFive -= 3;
                }
                if (numFive < 0 || numTen < 0) return false;
            }
        }
        return true;
    }
};

TEST(Test860, SimpleTest) {
    vector<int> bills{
        5,5,5,10,20
    };
    Solution s; 
    EXPECT_TRUE(s.lemonadeChange(bills));
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
